### ISSUE TITLE
Renderable Texture View must be "2d" and arrayLayerCount must be 1

### ISSUE THREAD

**Issue by MalekiRe** [gpuweb/gpuweb#4109]
The specification for what makes a GPUTextureView a renderable texture view includes that 
`descriptor.dimension` must be "2d" and
`descriptor.arrayLayerCount" must be 1.

This prevents the use of a multiview rendering for OpenXR where a "2darray" texture dimension and an arrayLayerCount of 2 can be used to create a view to render to each eye of a VR headset.

An example of which can be seen in this fork 
https://github.com/MalekiRe/wgpu-openxr-example

Is this a necessary restriction or can it be expanded to allow "2darray" and arrayLayerCount of 2?

**Comment by kainino0x** [comment]
A bunch more is needed to achieve multi-view rendering than just lifting that validation requirement. I've retitled accordingly. Thanks for the request!

**Comment by rektide** [comment]
https://github.com/immersive-web/WebXR-WebGPU-Binding has some overlap with this topic. Godspeed.

**Comment by magcius** [comment]
I think this can be considered a duplicate of these later bug reports?

https://github.com/gpuweb/gpuweb/issues/4812
https://github.com/gpuweb/gpuweb/issues/4806

**Comment by kainino0x** [comment]
I was thinking about like, `OVR_multiview`. @toji (and you) would know better if that concept is covered by those other issues.

**Comment by martinlaxenaire** [comment]
Hello folks,

Do we if know multi-view is still in the pipe and if we might someday be able to use it?
I know its main usage would be for WebXR, but it would help to be able to write to a cubemap in one pass for example (ref: https://kidrigger.dev/post/vulkan-render-to-cubemap-using-multiview/), especially for point shadows.

Thanks,

**Comment by alienself** [comment]
Any progress on multiview @toji? Specially now that we have WebXR support for WebGPU in Chrome, adding multiview would be a natural next step.

**Comment by toji** [comment]
I haven't been able to look into it personally yet, no. I do agree that it would be a great compliment to WebXR and would like to investigate in the near future.

**Comment by alienself** [comment]
Sounds good! Thank you for the update, definitely looking forward to it!

**Comment by Schlabbermampf** [comment]
Bumping for Multi-View support.
Currently it is possible to do Instanced-Stereo rendering by using a double-sized Render-Target, but then hardware-clipping is the problem: fragments for the left eye get drawn onto the right eye and vice versa. So you still need to discard pixels in the fragment. There seems to be a solution in OpenGL: https://stackoverflow.com/questions/29507534/instanced-based-vr-rendering-in-opengl

But a dedicated Multi-View solution would be nicer anyway i assume and getting the performance boost in VR might be crucial.

**Comment by magcius** [comment]
> Currently it is possible to do Instanced-Stereo rendering by using a double-sized Render-Target, but then hardware-clipping is the problem: fragments for the left eye get drawn onto the right eye and vice versa.

Use viewport/scissor to clip the drawn area?

**Comment by Schlabbermampf** [comment]
> Use viewport/scissor to clip the drawn area?

the scissor feature will not work because it applies per-pass but it needs to apply per-instance. 

But actually in WebGPU there is a builtin variable to solve the issue just like the OpenGL solution: 
@builtin(clip_distances) ClipDistances: array<f32, 2>

the feature is not available on all platforms though. here are some tests (it still might be different, depending on OS/Browser versions):
Windows native -> yes
Windows Chrome -> yes
Windows Firefox -> no
Android native -> no
Android Chrome -> no
Android Quest2 native -> (yes)

Though on the Quest2 there seems to be a driver issue: the shader does throw an VK_ERROR_UNKNOWN when creating the Render Pipeline. So I actually wasn't able yet to see the clip_distances feature in action.

Anyway this could be a nice solution to do Instanced-Stereo, but actual Multi-View support would still be a much cleaner approach (and i assume that Instancing generally has a different use-case).


Edit: When comparing the naive two-pass approach with the naive Instanced-Stereo variant (using discard in the fragment-shader, not clip_distance), we get a benefit of ~10 FPS on the Quest2, that's a motivation.

**Comment by cabanier** [comment]
I've been running some experiments and I think we only need minimal changes to enable multiview.
As an experiment, I added viewCount to WebGPU:GPURenderPassDescriptor:

```
dictionary GPURenderPassDescriptor : GPUObjectDescriptorBase {
    ...
    GPUSize32 viewCount;
};
```
This tells the GPU how many views to render in a single pass. When set, all color/depth attachments must be texture arrays with at least viewCount layers.

```
@builtin(view_index) in WGSL — add as a recognized builtin:
```

- Available in vertex and fragment shaders as an input
- Returns which view (eye) is currently being rendered
- The GPU provides it independently to each stage (not passed between stages)

Validation rules:
- When viewCount is set, all attachments must be array textures with arrayLayerCount >= viewCount
- Pipelines created for multiview passes are only compatible with render passes that have the same viewCount

A multiview example could look like this:
```
// Request device with multiview support
const adapter = await navigator.gpu.requestAdapter();
const device = await adapter.requestDevice({
  requiredFeatures: ["chromium-experimental-multiview"],
});

// Create a 2-layer texture array for multiview rendering
const texture = device.createTexture({
  size: [512, 512, 2], // 2 array layers, one per view
  format: "rgba8unorm",
  usage: GPUTextureUsage.RENDER_ATTACHMENT,
});

const shaderModule = device.createShaderModule({
  code: `
    struct VertexOutput {
      @builtin(position) position: vec4f,
      @location(0) color: vec3f,
    }

    @vertex
    fn vs(@builtin(vertex_index) vi: u32, @builtin(view_index) view: u32) -> VertexOutput {
      // Simple triangle
      var pos = array<vec2f, 3>(
        vec2f( 0.0,  0.5),
        vec2f(-0.5, -0.5),
        vec2f( 0.5, -0.5),
      );

      var out: VertexOutput;
      out.position = vec4f(pos[vi], 0.0, 1.0);

      // Different color per view
      if (view == 0u) {
        out.color = vec3f(1.0, 0.0, 0.0); // red for view 0
      } else {
        out.color = vec3f(0.0, 0.0, 1.0); // blue for view 1
      }
      return out;
    }

    @fragment
    fn fs(in: VertexOutput) -> @location(0) vec4f {
      return vec4f(in.color, 1.0);
    }
  `,
});

const pipeline = device.createRenderPipeline({
  layout: "auto",
  vertex: { module: shaderModule },
  fragment: {
    module: shaderModule,
    targets: [{ format: "rgba8unorm" }],
  },
});

// Render pass with viewCount: 2
const encoder = device.createCommandEncoder();
const pass = encoder.beginRenderPass({
  colorAttachments: [
    {
      view: texture.createView(), // full array view (both layers)
      loadOp: "clear",
      storeOp: "store",
    },
  ],
  viewCount: 2, // render both views in a single pass
});

pass.setPipeline(pipeline);
pass.draw(3);
pass.end();

device.queue.submit([encoder.finish()]);
// Layer 0 now has a red triangle, layer 1 has a blue triangle
```

I updated Dawn, WebGPU's implementation in Chromium and three.js to use this and am happy to share this code with the group.

**Comment by toji** [comment]
Thanks for copying over here, Rik! I'll repeat my comment from the other thread too:

Your outline of the multiview feature above looks like what I would expect, with the exception of you noting that "Pipelines created for multiview passes are only compatible with render passes that have the same viewCount", but the code example doesn't show any view count being set for the pipeline, and as far as I can tell at least [Vulkan](https://docs.vulkan.org/refpages/latest/refpages/source/VK_KHR_multiview.html) only requires it to be set for the render pass. Did you find that setting the view count on the pipeline was necessary?

**Comment by cabanier** [comment]
Sorry, I had to do so on the Vulkan level but not at the WebGPU level. Instead, I had Dawn handle that so the developer doesn't need to do so.

**Comment by toji** [comment]
How did you create the pipelines so that they had the view count in advance?

**Comment by cabanier** [comment]
I look at the shader metadata

**Comment by mwyrzykowski** [comment]
Should we break out vertex amplification into a separate issue?

Presumably that could be implemented on backends 

[Metal has it](https://developer.apple.com/documentation/metal/improving-rendering-performance-with-vertex-amplification?language=objc) as does [Direct3D I think](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_view_instancing_desc) as well as [Vulkan](https://docs.vulkan.org/refpages/latest/refpages/source/VK_KHR_multiview.html)

I think its optionally supported though, perhaps it should be an API + WGSL extension to avoid needing to polyfill it?

**Comment by cabanier** [comment]
> Should we break out vertex amplification into a separate issue?

@mwyrzykowski What is the difference between vertex amplification and multiview? Your link to Vulkan points to the multiview extension

**Comment by AlbinBernhardssonARM** [comment]
Multiview support is ubiquitous in Vulkan (99%+ according to [vulkan.gpuinfo.org](https://vulkan.gpuinfo.org/listfeaturescore11.php?platform=all)). Unless you need it for geometry/tessellation but let's not... [At least 6 views is supported everywhere.](https://vulkan.gpuinfo.org/displaycoreproperty.php?core=1.1&name=maxMultiviewViewCount&platform=all) So on Vulkan, we should be able to expose 6 views without needing to polyfill anything (as long as we can deal with [`maxMultiviewInstanceIndex`](https://vulkan.gpuinfo.org/displaycoreproperty.php?core=1.1&name=maxMultiviewInstanceIndex&platform=all) + if there are any driver bugs).

**Comment by Kangz** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2026%E2%80%9004%E2%80%9001#vertex-amplification---multi-view-rendering-4109">GPU Web WG 2026-04-01 Atlantic-time</a></summary>

* Mike's [latest comment](https://github.com/gpuweb/gpuweb/issues/4109#issuecomment-4129579550) gives pointers about vertex amplification  
* CW: APIs have different ways to broadcast vertices to multiple of the render targets. Has funny interactions with instancing \- imposes a cap on the max instance index (probably because hardware pipes it through instancing).  
* MW: Good summary. Vertices can be processed once and sent to multiple outputs.  
* CW: Q for group: Do we expect developers using multiview, to use any of the features from \#4812? Or do they just always broadcast to all views?  
* BJ: For WebXR, your vertex amplification would usually be 2, and one for each view. But for other cases like shadow mapping, unsure? Likely constant, need to find out.  
* BJ: Also not familiar yet with how to do vertex shift with vertex amplification, will read up.  
* — later  
* BJ: Vertex amplification isn't necessary, just an optimization to reduce the number of times the vertex shader runs?  
* RC: Difference between multiview and this?  
* CW: Multiview groups both features. I think the main important part to people is vertex amplification.

</details>

**Comment by cabanier** [comment]
@Kangz What is the next step?

**Comment by Kangz** [comment]
To make progress on this we need:

 1. A document in the [proposals](https://github.com/gpuweb/gpuweb/tree/main/proposals) folder describing the new capabilities and the changes to validation rules (extras for copying over the per-target API investigation API etc). IMHO we can have a single one for both vertex amplification and layer selection. The group approve the PR adding the document (that should be easy, as it is a living doc as we discuss things).
 2. A prototype, landed or not in one of the implementations (in this case just wiring to a JS runtime like node or deno can be sufficient to test against CTS).
 3. [WebGPU CTS](https://github.com/gpuweb/cts) coverage gated on the tentative feature name that passes on the prototype implementation.

Given the low risk of blocking issues for this extension, doing everything in parallel is possible. These three components are needed because they let the WebGPU group be confident that the feature is implementable as we always discover at least minor issues during implementation and testing. After all that is landed, a spec PR can be done and should land quickly.

However at the moment AFAIK all implementers are dealing with a stream of high-priority issues so I don't expect progress coming from them too soon. If you or someone else want to push it forward, it looks like [wgpu already has a multiview feature](https://docs.rs/wgpu/latest/wgpu/struct.Features.html#associatedconstant.MULTIVIEW) that could be massaged to match what would become upstream, or Dawn/Tint could take in a prototype of the feature. (the CTS is often just as much work as implementation).

**Comment by cabanier** [comment]
> To make progress on this we need:
> 
> 1. A document in the [proposals](https://github.com/gpuweb/gpuweb/tree/main/proposals) folder describing the new capabilities and the changes to validation rules (extras for copying over the per-target API investigation API etc). IMHO we can have a single one for both vertex amplification and layer selection. The group approve the PR adding the document (that should be easy, as it is a living doc as we discuss things).
> 2. A prototype, landed or not in one of the implementations (in this case just wiring to a JS runtime like node or deno can be sufficient to test against CTS).
> 3. [WebGPU CTS](https://github.com/gpuweb/cts) coverage gated on the tentative feature name that passes on the prototype implementation.

I created a proposal here: https://github.com/gpuweb/gpuweb/pull/6253
Quest browser will also ship with this feature behind experimental flags.
I can look into [WebGPU CTS](https://github.com/gpuweb/cts) next.
